src/main/java/org/opensearch/security/resources/ResourceAccessHandler.java

```diff
@@ -0,0 +1,343 @@
+            return false;  // If the document doesn't exist, no permissions can be granted
+        }
+
+        if (isSharedWithEveryone(document)
+            || isOwnerOfResource(document, user.getName())
+            || isSharedWithEntity(document, EntityType.USERS, Set.of(user.getName()), scope)
+            || isSharedWithEntity(document, EntityType.ROLES, userRoles, scope)
+            || isSharedWithEntity(document, EntityType.BACKEND_ROLES, userBackendRoles, scope)) {
+            LOGGER.info("User {} has {} access to {}", user.getName(), scope, resourceId);
+            return true;
+        }
+
+        LOGGER.info("User {} does not have {} access to {} ", user.getName(), scope, resourceId);
+        return false;
+    }
+
+    /**
+     * Shares a resource with the specified users, roles, and backend roles.
+     * @param resourceId The resource ID to share.
+     * @param resourceIndex  The index where resource is store
+     * @param shareWith The users, roles, and backend roles as well as scope to share the resource with.
+     * @return The updated ResourceSharing document.
+     */
+    public ResourceSharing shareWith(String resourceId, String resourceIndex, ShareWith shareWith) {
+        if (areArgumentsInvalid(resourceId, resourceIndex, shareWith)) {
+            return null;
+        }
+        final User user = threadContext.getPersistent(ConfigConstants.OPENDISTRO_SECURITY_USER);
+        LOGGER.info("Sharing resource {} created by {} with {}", resourceId, user.getName(), shareWith.toString());
+
+        // check if user is admin, if yes the user has permission
+        boolean isAdmin = adminDNs.isAdmin(user);
+
+        return this.resourceSharingIndexHandler.updateResourceSharingInfo(resourceId, resourceIndex, user.getName(), shareWith, isAdmin);
+    }
+
+    /**
+     * Revokes access to a resource for the specified users, roles, and backend roles.
+     * @param resourceId The resource ID to revoke access from.
+     * @param resourceIndex  The index where resource is store
+     * @param revokeAccess The users, roles, and backend roles to revoke access for.
+     * @param scopes The permission scopes to revoke access for.
+     * @return The updated ResourceSharing document.
+     */
+    public ResourceSharing revokeAccess(
+        String resourceId,
+        String resourceIndex,
+        Map<EntityType, Set<String>> revokeAccess,
+        Set<String> scopes
+    ) {
+        if (areArgumentsInvalid(resourceId, resourceIndex, revokeAccess, scopes)) {
+            return null;
+        }
+        final User user = threadContext.getPersistent(ConfigConstants.OPENDISTRO_SECURITY_USER);
+        LOGGER.info("User {} revoking access to resource {} for {} for scopes {} ", user.getName(), resourceId, revokeAccess, scopes);
+
+        // check if user is admin, if yes the user has permission
+        boolean isAdmin = adminDNs.isAdmin(user);
```

Comment:
do we want to split this out into a utility function such as hasAdminLevelAcess? Maybe there are additional use cases that we want to support here, such as a "super" role, and not just based on the DN... wdyt?

Reply:
This here is specific to super-admin. All "super" roles will require user to have the resource shared with them with that scope.